### PR TITLE
tests for voting system

### PR DESCRIPTION
**Pull request by CloudedCat**

### CODE REVIEW

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 1]
```diff
@@ -0,0 +1,425 @@
+package main
+
+import (
+	//"bytes"
+	"encoding/json"
+	//"errors"
+	"fmt"
+	//"github.com/alexeyknyshev/gojsondiff"
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Remove unused imports

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+	PatchId uint32 `json:"patch_id,omitempty"`
+	UserId  uint32 `json:"user_id"`
+	Score   uint32 `json:"score"`
+}
+
+func voteCompare(t *testing.T, res, expected *VotePatch) bool {
+	success := true
+	if res.UserId != expected.UserId {
+		t.Error("expected UserId = ", expected.UserId, "got UserId = ", res.UserId)
+		success = false
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: `return false` is more clear in my opinion

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+	metrics, err := getSpaceMetrics(hCtx)
+	if err != nil {
+		t.Errorf("Failed to get space metric on start: %v", err)
+	}
+	defer checkSpaceMetrics(t, func() ([]byte, error) { return getSpaceMetrics(hCtx) }, metrics)
+
+	lastPatch, CpId := invokeTaranPatchFuncs(t, hCtx, requestJson)
+	if lastPatch == 0 {
+		return
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Looks like test will be treated as passing in this case. Is it expected behaviour?

Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Same here

Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Same here

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+	success := true
+	if res.UserId != expected.UserId {
+		t.Error("expected UserId = ", expected.UserId, "got UserId = ", res.UserId)
+		success = false
+	}
+	if res.Score != expected.Score {
+		t.Error("expected Score = ", expected.Score, "got Score = ", res.Score)
+		success = false
+	}
+	return success
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: `return true`

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+		} else if i != 4 { //don't check timestamp
+			if resPatch[i].(uint64) != vol.(uint64) {
+				t.Error("comparePatches: fields", fields[i], "don't match")
+			}
+		}
+	}
+}
+
+func invokeTaranPatchFuncs(t *testing.T, hCtx *HandlerContextStruct, requestJson []byte) (uint32, uint64) {
+	TaranResp, err := hCtx.Tnt().Call("cashpointProposePatch", []interface{}{requestJson})
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: First part of with function is already implemented as [handlerCashpointCreate](https://github.com/alexeyknyshev/android_cash_points/blob/master/server/src/github.com/alexeyknyshev/cpsrv/handlers_cashpoints.go#L255)
It does not mean that your solution is wrong, but wrapping handler allow us to test http handler as well as tarantool-side logic.

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+	Eur            *bool    `json:"eur,omitempty"`
+	CashIn         *bool    `json:"cash_in,omitempty"`
+}
+
+type TestPatchReq struct {
+	Data   PatchRequest `json:"data"`
+	UserId uint         `json:"user_id"`
+}
+
+func getPatchExampleNewCP() *PatchRequest {
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Why return by **ptr** instead by **value**?

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+	return &patchReq
+
+}
+
+func getPatchExampleExistCP() (*PatchRequest, string) {
+	patchReq := PatchRequest{
+		Id:     58552,
+		BankId: 2764,
+	}
+	exampleJson := "{\"schedule\":{},\"bank_id\":" + strconv.FormatUint(uint64(patchReq.BankId), 10) + "}"
```
Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Is there any reason for passing `schedule`?

Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: BTW, you can use [anonymous structs](https://talks.golang.org/2012/10things.slide#2)
Example: https://play.golang.org/p/WSsay8kpgo

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go (hunk 2 above)
Comment (CloudedCat) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: My version show all possible error. It don't stop after first failed compare

Comment (alexeyknyshev) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: ok

File: server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go (hunk 3 above)
Comment (CloudedCat) [inline server/src/github.com/alexeyknyshev/cpsrv/cpsrv_patch_test.go]: Yes, invokeTaranPatchFuncs contain Error call. Thx for notice, I have seen suspicious code in that function

**Pull request merged**